### ISSUE TITLE
Depth normalisation

### ISSUE THREAD

**Issue by adele-morrison** [willaguiar/ASC_and_heat_transport#31]
As discussed in the hackathon today, different points on the isobath have different max depths, due to the discrete horizontal grid. Currently we are just doing straight depth averaging for the correlation plots. This means that some bottom points with depth 800m and being averaged with other points at 800m that are actually 400m above the bottom (where max depth = 1200m). It may be better to normalise depth and then average, so we always average bottom points together etc.

@ongqingyee is going to check what difference this makes for Figure 2.

**Comment by taimoorsohail** [comment]
For reference - this is what the months CSHT and U_along look like int arms of their max depth: 

![Screenshot 2024-04-11 at 12 28 44 PM](https://github.com/willaguiar/ASC_and_heat_transport/assets/20695740/e00b6e82-5ec0-4dd3-a969-33d91f8ff978)

This data is pulled from 

```
# Load the along-slope velocity field
u_along = xr.open_mfdataset('/g/data/v45/wf4500/ASC_project_files/Binned_ASC_speed/OM2_IAF/Binned_Antarctic_slope_contour_1km_velocities*')
# Load CSHT field
CSHT_along = xr.open_mfdataset(\
    '/g/data/v45/wf4500/ASC_project_files/Cross_slope_heat_transport/OM2_IAF/daily_z/*')
    ```

**Comment by willaguiar** [comment]
Ok, there seem to be a big mismatch in the isobath between the CSHT and U ~-80.  Im gonna check the code that produced the CSHT to see if the same isobath is used

**Comment by adele-morrison** [comment]
I don't think I was following the discussion on this closely before. Why is the max depth of the CSHT and u_along different at each longitude? I thought we used the same isobath for both?

**Comment by taimoorsohail** [comment]
Yes, this is also a point of confusion for me. I've highlighted the files I
drew this monthly data from above, so Will if you could check out why they
are different that would be great!

I won't produce the normalised, monthly-averaged u_along and CSHT files
until this mismatch is resolved @Ellie Ong ***@***.***>.

**Comment by taimoorsohail** [comment]
@willaguiar The data isn't binned, it's the raw longitudes I believe

**Comment by willaguiar** [comment]
ok, I checked the isobath used in the [ASC code, line 291](https://github.com/willaguiar/ASC_and_heat_transport/blob/main/Python/Daily_along_slope_velocities_XYgrid-monthly_output.py) and in the [CSHT, line 260](https://github.com/willaguiar/ASC_and_heat_transport/blob/main/Python/Daily_cross_slope_heat_transport_IAF_from_offline.py) and they come from the same file.

I also plotted the vars in the dirs pointed above. Even tho they don't look identical, the binned variables look very similar
<img width="1489" alt="Screenshot 2024-04-11 at 3 23 33 PM" src="https://github.com/willaguiar/ASC_and_heat_transport/assets/70033934/f1340d19-d8d3-466f-910c-d7d9e109bb6e">



@taimoorsohail did you use a specific time slice for that plot? My plots above are averages, so perhaps they are missing some peak deep value happening in one time slice, or missing something else.

(currently checking the unbinned data)

**Comment by taimoorsohail** [comment]
Thanks @willaguiar! I just plotted the first month in the monthly binned data for CSHT and U_along. Shouldn't they have the same isobath definition regardless of if it is time-averaged or if it is instantaneous? Here's the code snippet that pulls the CSHT, does the monthly averaging, and plots the first month:

```
CSHT_along = xr.open_mfdataset(\
    '/g/data/v45/wf4500/ASC_project_files/Cross_slope_heat_transport/OM2_IAF/daily_z/*')
CSHT_along_monthly = CSHT_along.resample(time='1M').mean()
CSHT_months = ((CSHT_along_monthly.binned_cross_slope_heat_trans+CSHT_along_monthly.zonal_convergence)*0.08).rename({'st_ocean':'depth', 'lon_bin_midpoints':'lon'})

fig, axs = plt.subplots(nrows=1, ncols=2, figsize=(10,5))
axs = axs.ravel()
plt.subplots_adjust(hspace = 1.2, wspace=0.2)

CSHT_months.where(CSHT_months!=0).isel(time=0).plot(ax=axs[0], cmap = plt.cm.Greens)
axs[0].invert_yaxis()
axs[0].set_ylim(2000,0)
axs[0].set_title('CSHT')
u_along_da_monthly.isel(time=0).plot(ax=axs[1], cmap = plt.cm.Blues)
axs[1].invert_yaxis()
axs[1].set_ylim(2000,0)
axs[1].set_title('U_along')
plt.show()

**Comment by willaguiar** [comment]
ok, I was looking at the code for the Zonal Convergence. The ZC is calculated from the difference x-transports not on the 1000m isobath, but at the points adjacent to the edges of each 3deg bin. So I think what is happening here is that the depths at the edge of certain bins can be bigger than the depth of the points inside the bins, creating the deep ZC values. Does that make sense @adele-morrison ?

If so perhaps the best thing to do is to mask out any ZC point that do not have CSHT in it

**Comment by willaguiar** [comment]
I talked to @adele-morrison in person on that issue today. It seems reasonable that the ZC would have deeper values, as they are not taken from the exact same points as the Binned CSHT. The only problem that could arise is in the vertical profile analysis ( where values z>1200m would come from pure ZC), but since we are cutting of our analysis at ~1000m depth then that should not be a problem.

One thing to think about it tho is whether having these deep ZC-driven transports would affect the GMM classification ( probably not much?  - @taimoorsohail )

**Comment by taimoorsohail** [comment]
Hmmm, does that mean we aren't going to vertically normalise the CSHT and
U_along profiles anymore, and will instead cut them off at exactly 1000m? I
think as we are calculating correlations between U_along and CSHT+ZC, the
values of ZC that are deeper than ~1000m will not contribute to the
correlation at all (as U_along will be NaN at these points). So why don't
we just NaN out CSHT+ZC where the contribution is ZC only, then normalise
the resulting profiles of CSHT+ZC and U_along?

**Comment by adele-morrison** [comment]
Looking at the unbinned plot of CSHT above, it has a much larger range of max depths than the binned. I think it would still be good to test the normalisation, but we definitely want to use the max depths from the unbinned data to do that. The binning process doesn't change the depth of the CSHT values, it just possibly adds in deeper values from the ZC terms. So perhaps we could test throwing out any ZC data deeper than the max depth in the unbinned CSHT and normalising what's left that's shallower than the max depth of the unbinned CSHT?

**Comment by taimoorsohail** [comment]
OK - that sounds like a good plan. @willaguiar do you have the path to the unbinned CSHT and U_along (which you plotted above)? I will try to do the normalisation with the unbinned CSHT+ZC data, throwing out the values of ZC that are deeper than the max depth of CSHT at each longitude. 

Just confirming - does this mean the daily and monthly analyses should be done with unbinned data rather than the binned data? Shifting from binned to unbinned data (even if we cut off the correlation at 1000m, removing the influence of the deep ZC values) will have an impact on the correlations, right?

**Comment by taimoorsohail** [comment]
> One thing to think about it tho is whether having these deep ZC-driven transports would affect the GMM classification ( probably not much? - @taimoorsohail )

I classified with GMM using only the binned, monthly U_along values, so CSHT doesn't affect my classification at all.

**Comment by willaguiar** [comment]
I'm not sure I understand completely -  is the idea to rerun the normalization + GMM with the unbinned data? the only problem is that the ZC is calculated in the edges of the bins, and therefore only exist in the binned data points. So Im not sure how would be possible to filter the ZC to data points where the unbinned CSHT exists.

In any case, here it goes the dirs for the unbinned data:

unbinned ASC speed: `/g/data/v45/wf4500/ASC_project_files/ASC_speed/OM2_IAF/*`
CSHT should be the same files, but var: `unbinned_heat_transp_across_contour`

**Comment by adele-morrison** [comment]
Hmm, yeah, ok. What about finding the max depth for normalisation from the unbinned data in each bin? But then normalising the binned data. Could either find the mean isobath depth or max isobath depth in each bin?

**Comment by taimoorsohail** [comment]
But then we'd have a bunch of Nans at the bottom of the normalised profiles, which correspond to the deeper grid cells in the unbinned data that don't have any data in the binned array? If we are normalising the binned data at the end of the day, I don't see how incorporating the max depths from the unbinned CSHT will affect our result, other than adding a bunch of NaNs at the bottom of every profile... maybe I'm misunderstanding...?

**Comment by adele-morrison** [comment]
Hmm I'm not sure I understand, maybe we should zoom?

**Comment by adele-morrison** [comment]
I think we'd have to use the max isobath depth in each bin, then any data beneath that comes from the ZC and we just throw that out when normalising.

**Comment by taimoorsohail** [comment]
Yes let's chat over Zoom.

**Comment by taimoorsohail** [comment]
OK Adele and I just Zoomed about this. It seems there are two steps we need to accomplish. 

1) The binning process combines a series of adjacent, shallow CSHT profiles (see around -150 in the unbinned data). In the process of this binning, it may combine only one or two data points at the very bottom, adding noise to the solution. We need to try to avoid this by reducing the thickness of the bins in the longitude binning. However, we are concerned this will add noise, so we need to test how much of a difference this will make to the correlations. I will make an issue about this now. 
 
2) If we do reduce our bin size without major impacts on our correlations, I will then normalise the newly binned CSHT+ZC and U_along profiles, taking care to mask out the depths where CSHT=0. 

This will feed into the monthly analysis by @ongqingyee, the coarsening work by @fabiobdias and the daily analysis by @willaguiar [assuming this is all done using the binned data...?].
